external-crates/move/crates/move-package/src/lock_file/schema.rs

```diff
@@ -237,3 +246,66 @@ fn to_toml_edit_value(value: &toml::Value) -> toml_edit::Item {
         }
     }
 }
+
+pub enum ManagedAddressUpdate {
+    Published {
+        original_id: String,
+        chain_id: String,
+    },
+    Upgraded {
+        latest_id: String,
+        version: u64,
+    },
+}
+
+/// Saves published or upgraded package addresses in the lock file.
+pub fn update_managed_address(
+    file: &mut LockFile,
+    environment: String,
+    managed_address_update: ManagedAddressUpdate,
+) -> Result<()> {
+    use toml_edit::{value, Document, Item, Table};
+
+    let mut toml_string = String::new();
+    file.read_to_string(&mut toml_string)?;
+    let mut toml = toml_string.parse::<Document>()?;
+
+    let env_table = toml
+        .entry(ENV_TABLE_NAME)
+        .or_insert_with(|| Item::Table(Table::new()))
+        .as_table_mut()
+        .ok_or_else(|| anyhow!("Could not find or create 'env' table in Move.lock"))?
+        .entry(&environment)
+        .or_insert_with(|| Item::Table(Table::new()))
+        .as_table_mut()
+        .ok_or_else(|| anyhow!("Could not find or create {environment} table in Move.lock"))?;
+
+    match managed_address_update {
+        ManagedAddressUpdate::Published {
+            original_id,
+            chain_id,
+        } => {
+            env_table[CHAIN_ID_KEY] = value(chain_id);
+            env_table[ORIGINAL_PUBLISHED_ID_KEY] = value(original_id.to_string());
```

Comment:
nit: use of `to_string()` as a way to `clone()` is a little funky here -- could we just pass a reference to `value`, or failing that, use `clone` directly?